syntax = "proto3";

package nameNode;

service NameNodeService {
  // DataNodes
  rpc Register (RegisterRequest) returns (RegisterResponse);
  rpc GetDataNodesForUpload (DataNodesUploadRequest) returns (DataNodesResponse);
  rpc GetDataNodesForDownload (DataNodesDownloadRequest) returns (DataNodesDownloadResponse);

  // Users
  rpc AddUser (AddUserRequest) returns (AddUserResponse);
  rpc ValidateUser (ValidateUserRequest) returns (ValidateUserResponse);
}

message RegisterRequest {
  string ip = 1;
  string port = 2;
  double capacity_MB = 3;
}

message RegisterResponse {
  string id = 1;
}

message DataNodesUploadRequest {
  string file = 1;
  double size = 2;
  string username = 3;
}

message DataNodesResponse {
  repeated DataNodeInfo nodes = 1;
  string block_id = 2; 
}

message DataNodeInfo {
  string id = 1;
  string ip = 2;
  string port = 3;
  double capacity_MB = 4;
}

message AddUserRequest {
  string username = 1;
  string password = 2;
}

message AddUserResponse {
  string status = 1;
}

message ValidateUserRequest {
  string username = 1;
  string password = 2;
}

message ValidateUserResponse {
  string status = 1;
}

message DataNodesDownloadRequest {
  string file = 1;
  string username = 2;
}

message DataNodesDownloadResponse {
  map<int32, string> blocks = 1;
}
